
<a href="https://qworld.net" target="_blank" align="left"><img src="../qworld/images/header.jpg"  align="left"></a>
<table  align="left" width="100%"> <tr>
        <td  align="right" style="background-color:#ffffff;vertical-align:bottom;horizontal-align:right">
            prepared by Özlem Salehi (<a href="http://qworld.lu.lv/index.php/qturkey/" target="_blank">QTurkey</a>)
        </td>        
</tr></table>

<table width="100%"><tr><td style="color:#bbbbbb;background-color:#ffffff;font-size:11px;font-style:italic;text-align:right;">This cell contains some macros. If there is a problem with displaying mathematical formulas, please run this cell to load these macros. </td></tr></table>
$ \newcommand{\bra}[1]{\langle #1|} $
$ \newcommand{\ket}[1]{|#1\rangle} $
$ \newcommand{\braket}[2]{\langle #1|#2\rangle} $
$ \newcommand{\dot}[2]{ #1 \cdot #2} $
$ \newcommand{\biginner}[2]{\left\langle #1,#2\right\rangle} $
$ \newcommand{\mymatrix}[2]{\left( \begin{array}{#1} #2\end{array} \right)} $
$ \newcommand{\myvector}[1]{\mymatrix{c}{#1}} $
$ \newcommand{\myrvector}[1]{\mymatrix{r}{#1}} $
$ \newcommand{\mypar}[1]{\left( #1 \right)} $
$ \newcommand{\mybigpar}[1]{ \Big( #1 \Big)} $
$ \newcommand{\sqrttwo}{\frac{1}{\sqrt{2}}} $
$ \newcommand{\dsqrttwo}{\dfrac{1}{\sqrt{2}}} $
$ \newcommand{\onehalf}{\frac{1}{2}} $
$ \newcommand{\donehalf}{\dfrac{1}{2}} $
$ \newcommand{\hadamard}{ \mymatrix{rr}{ \sqrttwo & \sqrttwo \\ \sqrttwo & -\sqrttwo }} $
$ \newcommand{\vzero}{\myvector{1\\0}} $
$ \newcommand{\vone}{\myvector{0\\1}} $
$ \newcommand{\stateplus}{\myvector{ \sqrttwo \\  \sqrttwo } } $
$ \newcommand{\stateminus}{ \myrvector{ \sqrttwo \\ -\sqrttwo } } $
$ \newcommand{\myarray}[2]{ \begin{array}{#1}#2\end{array}} $
$ \newcommand{\X}{ \mymatrix{cc}{0 & 1 \\ 1 & 0}  } $
$ \newcommand{\Z}{ \mymatrix{rr}{1 & 0 \\ 0 & -1}  } $
$ \newcommand{\Htwo}{ \mymatrix{rrrr}{ \frac{1}{2} & \frac{1}{2} & \frac{1}{2} & \frac{1}{2} \\ \frac{1}{2} & -\frac{1}{2} & \frac{1}{2} & -\frac{1}{2} \\ \frac{1}{2} & \frac{1}{2} & -\frac{1}{2} & -\frac{1}{2} \\ \frac{1}{2} & -\frac{1}{2} & -\frac{1}{2} & \frac{1}{2} } } $
$ \newcommand{\CNOT}{ \mymatrix{cccc}{1 & 0 & 0 & 0 \\ 0 & 1 & 0 & 0 \\ 0 & 0 & 0 & 1 \\ 0 & 0 & 1 & 0} } $
$ \newcommand{\norm}[1]{ \left\lVert #1 \right\rVert } $
$ \newcommand{\pstate}[1]{ \lceil \mspace{-1mu} #1 \mspace{-1.5mu} \rfloor } $

<h1> <font color="blue"> Solutions for </font> Phase Estimation</h1>

<h3>Task 1 (on paper)</h3>
<a id="task1"></a>

Show that $\ket{-}$ and $\ket{+}$ are eigenvectors for the $X$ operator with eigenvalues -1 and 1 respectively.

<h3>Solution </h3>



\begin{align*}
X \ket{-} = X \frac {\ket{0} - \ket{1}} {\sqrt{2}}
& =  \frac {\ket{1} - \ket{0}} {\sqrt{2}}\\
& = -\frac {\ket{0} - \ket{1}} {\sqrt{2}}\\
& = -\ket{-}
\end{align*}


\begin{align*}
X \ket{+} = X \frac {\ket{0} + \ket{1}} {\sqrt{2}}
& =  \frac {\ket{1} + \ket{0}} {\sqrt{2}}\\
& = \frac {\ket{0} + \ket{1}} {\sqrt{2}}\\
& = \ket{+}
\end{align*}

Hence $ \ket{-}$ and $\ket{+}$ are eigenvectors of $ X $ operator with the eigenvalues -1 and 1 respectively.

<h3>Task 2 (on paper)</h3>
<a id="task2"></a>

Consider the following quantum state where $ x=0 $ or $ 1 $. How can you find out the value of $x$ by applying a single operator?

$$ 
\frac {\ket{0} + (-1)^x \ket{1}} {\sqrt{2}}
$$


<h3>Solution </h3>

If our aim is to find out $ x $, then applying a Hadamard gate we can determine whether $ x=0 $ or $ 1 $ depending on the outcome. If the outcome is $ \ket{0} $, then $ x=0 $ and if the outcome is $ \ket{1} $, then $ x=1 $.

<a id="task3"></a>
<h3>Task 3</h3>

Consider the unitary operator $U$ with eigenvector $\ket{1}$ and eigenvalue $e^{2\pi i \phi}$ where $\phi = \frac{5}{16}$. In this particular example, $U$ can be obtained with a Phase gate.

In [8]:
from qiskit.circuit.library import PhaseGate
from qiskit.quantum_info.operators import Operator
from math import pi

phase = 5/16
U = Operator(PhaseGate(2*pi*phase)) #We need a 2*pi since CP adds a phase of e^{i \phi}
print(U)

Operator([[ 1.        +0.j        ,  0.        +0.j        ],
          [ 0.        +0.j        , -0.38268343+0.92387953j]],
         input_dims=(2,), output_dims=(2,))


Pick $t=4$ and implement the phase estimation circuit to estimate $\phi$. Write a function named <i>qpe</i> which takes as parameters <i>t</i>, <i>control</i> qubits, <i>target</i> qubits, <i>circuit</i> and the operator <i>U</i>.

Note that you will get an exact result since $t=4$ is precise enough.

<i>Remarks:</i>

- You can apply arithmetical operations to gates, such as taking powers or inverses.
- To apply $QFT^{\dagger}$, you can use the function you have written in the previous notebook.


<h3>Solution </h3>

In [15]:
from qiskit import QuantumRegister, ClassicalRegister, QuantumCircuit
from qiskit_aer import AerSimulator
from math import pi

def iqft(n, qubits, circuit):
    #Swap the qubits
    for i in range(n//2):
        circuit.swap(qubits[i],qubits[n-i-1])     
        
    #For each qubit
    for i in range(n-1,-1,-1):
        #Apply CR_k gates where j is the control and i is the target
        k=n-i #We start with k=n-i
        for j in range(n-1,i,-1):
            #Apply CR_k gate  
            circuit.cp(-pi*2/2**(k), qubits[j],qubits[i])
            k=k-1 #Decrement k at each step
 
        #Apply Hadamard to the qubit
        circuit.h(qubits[i])

In [16]:
#%%writefile qpe.py
from qiskit import QuantumRegister, ClassicalRegister, QuantumCircuit
from qiskit_aer import AerSimulator
from qiskit.circuit.library import UnitaryGate

def qpe(t,control, target, circuit, U):
    
    #Apply Hadamard to control qubits
    for qb in control:
        circuit.h(qb)
    
    # Apply controlled unitaries
    for i in range(t):
        # Assemble list of control and target qubits
        qubits=[control[t-i-1]]
        for qq in target:
            qubits.append(qq)

        # Apply CU gates
        circuit.append(UnitaryGate(U).control(),qubits)

        # Increase the power of U for next iteration
        U=Operator(UnitaryGate(U.dot(U)))
       
    #Apply inverse QFT
    iqft(t,control,circuit)

In [17]:

from qiskit import QuantumRegister, ClassicalRegister, QuantumCircuit, transpile
from qiskit_aer import AerSimulator

#Set n and t
t=4 #Number of qubits in the control register
n=1 #Number of qubits in the register storing eigenvector

#Create t qubits of the control register
control = QuantumRegister(t)

#Create n qubits of the target register
target = QuantumRegister(n)

#Create classical register to measure control qubits
measurement = ClassicalRegister(t)

#Create cirucit
qc1=QuantumCircuit(control, target, measurement) 

#Set target qubit to state |1>
qc1.x(target[0])

#Apply QPE
qpe(t,control, target, qc1, U)

# Measure control qubits
qc1.measure(control, measurement)


# Reverse target (and every other) qubits
qc1=qc1.reverse_bits()
qc1_transpiled = transpile(qc1, AerSimulator())
job = AerSimulator().run(qc1_transpiled,shots=1000)
counts = job.result().get_counts(qc1_transpiled)
print(counts)                        


{'0101': 1000}


We obtain 0101 which is 5. As $t=4$, we get the phase $5/16$.

<a id="task4"></a>
<h3>Task 4</h3>

Try Task 3 this time by setting $t=3$. What do you expect to see?

<h3>Solution </h3>

In [ ]:
from qiskit import QuantumRegister, ClassicalRegister, QuantumCircuit, transpile
from qiskit_aer import AerSimulator
from qiskit.visualization import plot_histogram

t=3 #Number of qubits in the control register
n=1 #Number of qubits in the register storing eigenvector

#Create t qubits of the control register
control2 = QuantumRegister(t)

#Create n qubits of the target register
target2 = QuantumRegister(n)

#Create classical circuit to measure control qubits
measurement2 = ClassicalRegister(t)

#Create cirucit
qc2=QuantumCircuit(control2, target2, measurement2) 

#Set target qubit to state |1>
qc2.x(target2[0])

#Apply QPE
qpe(t,control2, target2, qc2, U)

# Measure control qubits
qc2.measure(control2, measurement2)

qc2=qc2.reverse_bits()
qc2_transpiled = transpile(qc2, AerSimulator())
job = AerSimulator().run(qc2_transpiled,shots=1000)
counts = job.result().get_counts(qc2_transpiled)
display(plot_histogram(counts))
print(counts)


The most probable outcomes are 011 and 010 corresponding to 2 and 3. Those would yield the phase estimates $2/8$ and $3/8$.

<a id="task5"></a>
<h3>Task 5</h3>

Try Task 3 this time by setting $\phi=0.685$. Calculate and print the estimates for different $t$ values between 1 and 10. 

<h3>Solution </h3>

In [ ]:
from qiskit.circuit.library import PhaseGate
from qiskit.quantum_info.operators import Operator
from math import pi

phase = 0.685
U=Operator(PhaseGate(2*pi*phase))

In [ ]:
from qiskit import QuantumRegister, ClassicalRegister, QuantumCircuit, transpile
from qiskit_aer import AerSimulator
from qiskit.visualization import plot_histogram

n=1 #Number of qubits in the register storing eigenvector

for t in range (1,10):

    #Create t qubits of the control register
    control3 = QuantumRegister(t)

    #Create n qubits of the target register
    target3 = QuantumRegister(n)

    #Create classical circuit to measure control qubits
    measurement3 = ClassicalRegister(t)

    #Create cirucit
    qc3=QuantumCircuit(control3, target3, measurement3) 

    #Set target qubit to state |1>
    qc3.x(target3[0])

    #Apply QPE
    qpe(t,control3, target3, qc3, U)

    # Measure control qubits
    qc3.measure(control3, measurement3)

    qc3=qc3.reverse_bits()
    qc3_transpiled = transpile(qc3, AerSimulator())
    job = AerSimulator().run(qc3_transpiled,shots=1000)
    counts = job.result().get_counts(qc3_transpiled)
    #display(plot_histogram(counts))

    # Most frequent observation
    freq=int(max(counts.keys(), key=(lambda new_k: counts[new_k])), 2)
    print("t:", t,"estimation:", freq/2**t)
    

In [ ]:
max(counts.keys(), key=(lambda new_k: counts[new_k]))
max(counts)

<a id="task6"></a>
<h3>Task 6</h3>

You are given an operator $U$ (Qiskit Operator object) with eigenvector $\ket{11}$ and eigenvalue $\phi$, which is this time unknown to you. Use phase estimation to estimate $\phi$ by trying different $t$ values. 

Run the following cell to load operator $U$

In [ ]:
%run estimate_operator.py

<h3>Solution </h3>

In [ ]:
from qiskit import QuantumRegister, ClassicalRegister, QuantumCircuit
from qiskit_aer import AerSimulator
from qiskit.visualization import plot_histogram

n=2 #Number of qubits in the register storing eigenvector

for t in range (1,10):

    #Create t qubits of the control register
    control4 = QuantumRegister(t)

    #Create n qubits of the target register
    target4 = QuantumRegister(n)

    #Create classical circuit to measure control qubits
    measurement4 = ClassicalRegister(t)

    #Create cirucit
    qc4=QuantumCircuit(control4, target4, measurement4) 

    #Set target qubit to state |1>
    qc4.x(target4[0])
    qc4.x(target4[1])

    #Apply QPE
    qpe(t,control4, target4, qc4, U)

    # Measure control qubits
    qc4.measure(control4, measurement4)

    qc4=qc4.reverse_bits()
    qc4_transpiled = transpile(qc4, AerSimulator())
    job = AerSimulator().run(qc4_transpiled,shots=1000)
    counts = job.result().get_counts(qc4_transpiled)
#   display(plot_histogram(counts))
  
    # Most frequent observation
    freq=int(max(counts.keys(), key=(lambda new_k: counts[new_k])), 2)
    print("t:", t,"estimation:", freq/2**t)
    